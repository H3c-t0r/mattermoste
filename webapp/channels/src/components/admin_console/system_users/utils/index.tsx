// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react';
import {FormattedMessage} from 'react-intl';

import {UserReportSortColumns, ReportSortDirection} from '@mattermost/types/reports';
import type {UserReportOptions, UserReport} from '@mattermost/types/reports';

import {PAGE_SIZES, type SortingState} from 'components/admin_console/list_table';

import type {AdminConsoleUserManagementTableProperties} from 'types/store/views';

import {ColumnNames, RoleFilters, StatusFilter} from '../constants';
import type {TableOptions} from '../system_users';

export function convertTableOptionsToUserReportOptions(tableOptions?: TableOptions): UserReportOptions {
    return {
        page_size: tableOptions?.pageSize || PAGE_SIZES[0],
        from_column_value: tableOptions?.fromColumnValue,
        from_id: tableOptions?.fromId,
        direction: tableOptions?.direction,
        ...getSortColumnForOptions(tableOptions?.sortColumn),
        ...getSortDirectionForOptions(tableOptions?.sortIsDescending),
        search_term: tableOptions?.searchTerm,
        ...getUserStatusFilterOption(tableOptions?.filterStatus),
        ...getRoleFilterOption(tableOptions?.filterRole),
    };
}

/**
 * Converts the sorting column name to API compatible sorting column name. Default sorting column name is by username.
 */
export function getSortColumnForOptions(id?: SortingState[0]['id']): Pick<UserReportOptions, 'sort_column'> {
    let sortColumn: UserReportOptions['sort_column'];

    if (id === ColumnNames.email) {
        sortColumn = UserReportSortColumns.email;
    } else if (id === ColumnNames.createAt) {
        sortColumn = UserReportSortColumns.createAt;
    } else {
        // Default sorting to first User details column
        sortColumn = UserReportSortColumns.username;
    }

    return {
        sort_column: sortColumn,
    };
}

/**
 * Converts the sorting direction to API compatible sorting direction. Default sorting direction is ascending.
 */
export function getSortDirectionForOptions(desc?: SortingState[0]['desc']): Pick<UserReportOptions, 'sort_direction'> {
    let sortDirection: UserReportOptions['sort_direction'];

    if (desc) {
        sortDirection = ReportSortDirection.descending;
    } else {
        sortDirection = ReportSortDirection.ascending;
    }

    return {
        sort_direction: sortDirection,
    };
}

/**
 * It returns the value of that column on which sorting is applied.
 */
export function getSortableColumnValueBySortColumn(row: UserReport, sortColumn: AdminConsoleUserManagementTableProperties['sortColumn']): string {
    switch (sortColumn) {
    case ColumnNames.email:
        return row.email;
    case ColumnNames.createAt:
        return String(row.create_at);
    default:
        return row.username;
    }
}

export function getUserStatusFilterOption(status?: string): Partial<Pick<UserReportOptions, 'hide_active' | 'hide_inactive'>> {
    if (status === StatusFilter.Active) {
        return {
            hide_inactive: true,
        };
    } else if (status === StatusFilter.Deactivated) {
        return {
            hide_active: true,
        };
    }

    return {};
}

export function getPaginationInfo(pageIndex: number, pageSize: number, currentLength: number, total?: number) {
    if (!currentLength) {
        return (
            <FormattedMessage
                id='admin.system_users_list.pagination.no_users'
                defaultMessage='0 users'
            />
        );
    }

    const firstPage = (pageIndex * pageSize) + 1;
    const lastPage = (pageIndex * pageSize) + currentLength;
    const totalItems = total || 0;

    return (
        <FormattedMessage
            id='admin.system_users_list.pagination'
            defaultMessage='Showing {firstPage} - {lastPage} of {totalItems} users'
            values={{
                firstPage,
                lastPage,
                totalItems,
            }}
        />
    );
}

export function getDefaultValueFromList<T extends {value: string}>(value: string, options: T[]) {
    const option = options.find((option) => option.value === value);

    if (option) {
        return option;
    }

    return options[0];
}

export function getRoleFilterOption(role?: string): Pick<UserReportOptions, 'role_filter'> {
    if (!role || role === RoleFilters.Any) {
        return {role_filter: ''};
    }
    return {role_filter: role};
}
