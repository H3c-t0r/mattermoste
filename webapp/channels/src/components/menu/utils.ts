// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {Children, ReactNode, cloneElement, isValidElement} from 'react';

import {MENU_ITEM_KEY_PREFIX, Props as MenuItemProps} from './menu_item';
import {SUB_MENU_ITEM_KEY_PREFIX} from './sub_menu';

export function injectPropsInMenuItems(children: ReactNode[], forceCloseMenu?: MenuItemProps['forceCloseMenu'], forceCloseSubMenu?: MenuItemProps['forceCloseSubMenu']) {
    const modifiedChildren = Children.map(children, (child) => {
        if (!isValidElement(child)) {
            return null;
        }

        if (
            child.props &&
            child.props.id &&
            (child.props.id.startsWith(MENU_ITEM_KEY_PREFIX) || child.props.id.startsWith(SUB_MENU_ITEM_KEY_PREFIX))
        ) {
            return cloneElement(child, {
                forceCloseMenu,
                forceCloseSubMenu,
            } as MenuItemProps);
        }

        return child;
    });

    return modifiedChildren;
}

export function createMenusUniqueId(prefix: string, menuItemName: string, ...uniqueValues: string[]) {
    const uniqueValuesJoined = uniqueValues.join('-');
    return `${prefix}__${menuItemName}__${uniqueValuesJoined}`;
}
