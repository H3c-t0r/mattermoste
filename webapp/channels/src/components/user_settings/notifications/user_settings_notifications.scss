.customKeywordsWithNotificationSubsection {
    & .multiInput {
        margin-block-start: 10px;
    }
}

.notificationSettingsModalHeader {
    display: flex;

    > span {
        align-self: center;
        margin-left: auto;

        a {
            display: flex;

            &:hover,
            &:active,
            &:focus {
                color: var(--button-bg);
                text-decoration: none;
            }

            > i {
                font-size: 14.4px;
            }

            > span {
                font-size: 12px;
                font-weight: 600;
                line-height: 16px;

                &:hover {
                    text-decoration: underline;
                }
            }
        }
    }

    .icon--in-a-circle {
        background: rgba(var(--button-bg-rgb), 0.12);
        width: 18px;
        height: 18px;
        margin-right: 4px;
        text-align: center;
        padding-left: 1px;
        border-radius: 50%;
    }
}
