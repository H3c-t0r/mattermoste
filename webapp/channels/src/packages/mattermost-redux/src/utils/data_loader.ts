// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

/**
 * A DataLoader is an object that can be used to batch requests for fetching objects from the server for performance
 * reasons.
 */
abstract class DataLoader<Identifier, Result = unknown> {
    protected readonly fetchBatch: (identifiers: Identifier[]) => Result;
    private readonly maxBatchSize: number;

    protected readonly pendingIdentifiers = new Set<Identifier>();

    constructor(args: {
        fetchBatch: (identifiers: Identifier[]) => Result;
        maxBatchSize: number;
    }) {
        this.fetchBatch = args.fetchBatch;
        this.maxBatchSize = args.maxBatchSize;
    }

    public addIdsToLoad(identifiersToLoad: Identifier[]) {
        for (const identifier of identifiersToLoad) {
            if (!identifier) {
                continue;
            }

            this.pendingIdentifiers.add(identifier);
        }
    }

    public doFetchBatch(): Result {
        let identifiersToLoad;

        // Since we can only fetch a defined number of user statuses at a time, we need to batch the requests
        if (this.pendingIdentifiers.size >= this.maxBatchSize) {
            identifiersToLoad = [];

            // We use temp buffer here to store up until max buffer size
            // and clear out processed user ids
            for (const identifier of this.pendingIdentifiers) {
                identifiersToLoad.push(identifier);
                this.pendingIdentifiers.delete(identifier);

                if (identifiersToLoad.length >= this.maxBatchSize) {
                    break;
                }
            }
        } else {
            // If we have less than max buffer size, we can directly fetch the statuses
            identifiersToLoad = Array.from(this.pendingIdentifiers);
            this.pendingIdentifiers.clear();
        }

        return this.fetchBatch(identifiersToLoad);
    }
}

/**
 * An IntervalDataLoader is an object that can be used to batch requests for fetching objects from the server. Instead
 * of requesting data immediately, it will periodically check if any objects need to be requested from the server.
 *
 * It's intended to be used for loading low priority data such as information needed in response to WebSocket messages
 * that the user won't see immediately.
 */
export class IntervalDataLoader<Identifier, Result = unknown> extends DataLoader<Identifier, Result> {
    private intervalId: number = -1;

    private doFetchBatchIfNecessary() {
        if (this.pendingIdentifiers.size === 0) {
            return;
        }

        this.doFetchBatch();
    }

    startIntervalIfNeeded(ms: number) {
        if (this.intervalId !== -1) {
            return;
        }

        this.intervalId = window.setInterval(() => this.doFetchBatchIfNecessary(), ms);
    }

    stopInterval() {
        clearInterval(this.intervalId);
        this.intervalId = -1;
    }
}

/**
 * A DelayedDataLoader is an object that can be used to batch requests for fetching objects from the server. Instead of
 * requesting data immediately, it will wait for an amount of time and then send a request to the server for all of
 * the data which would've been requested during that time.
 *
 * More specifically, when addIdsToLoad is first called, a timer will be started. Until that timer expires, any other
 * calls to addIdsToLoad will have the provided identifiers added to the ones from the initial call. When the timer
 * finally expires, the request will be sent to the server to fetch that data. After that, the timer will be reset and
 * the next call to addIdsToLoad will start a new one.
 *
 * DelayedDataLoader is intended to be used for loading data for components which are unaware of each other and may appear
 * in different places in the UI from each other which could otherwise send repeated requests for the same or similar
 * data as one another.
 */
export class DelayedDataLoader<Identifier, Result> extends DataLoader<Identifier, Promise<Result>> {
    private readonly wait: number = -1;

    private timeoutId: number = -1;
    private timeoutCallbacks = new Set<(result: Result) => void>();

    constructor(args: {
        fetchBatch: (identifiers: Identifier[]) => Promise<Result>;
        maxBatchSize: number;
        wait: number;
    }) {
        super(args);

        this.wait = args.wait;
    }

    public addIdsToLoad(identifiersToLoad: Identifier[]): Promise<Result> {
        return new Promise((resolve) => {
            super.addIdsToLoad(identifiersToLoad);

            // Save the callback that will resolve this promise so that the caller of this method can wait for its
            // data to be loaded
            this.timeoutCallbacks.add(resolve);

            this.startTimeoutIfNeeded();
        });
    }

    private startTimeoutIfNeeded() {
        if (this.timeoutId !== -1) {
            return;
        }

        this.timeoutId = window.setTimeout(() => {
            // Ensure that timeoutId and timeoutCallbacks are cleared before doing anything async so that any calls to
            // addIdsToLoad which come while we're fetching this batch are added to the next batch instead
            this.timeoutId = -1;

            const callbacks = Array.from(this.timeoutCallbacks);
            this.timeoutCallbacks.clear();

            // TODO this doesn't handle when someone requests more than a page of data
            this.doFetchBatch().then((result) => {
                // Unblock everyone waiting on this batch of data
                for (const callback of callbacks) {
                    callback(result);
                }
            });
        }, this.wait);
    }
}
