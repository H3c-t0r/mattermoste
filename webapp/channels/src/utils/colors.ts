// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

// Values taken from https://www.w3.org/TR/WCAG22/#dfn-relative-luminance
const COEFFICIENTS_OF_CIE_COLOR_SPACE = {
    RED: 0.2126,
    GREEN: 0.7152,
    BLUE: 0.0722,
};
const GAMMA_CORRECTION = 2.4;

// See https://www.w3.org/TR/2008/REC-WCAG20-20081211/#relativeluminancedef
function calculateRelativeLuminance(red: number, green: number, blue: number) {
    const gammaCorrectedScaledColors = [red, green, blue].map((color) => {
        const scaledColor = color / 255;

        if (scaledColor <= 0.03928) {
            return scaledColor / 12.92;
        }

        return Math.pow((scaledColor + 0.055) / 1.055, GAMMA_CORRECTION);
    });

    return sumRGBContributions(gammaCorrectedScaledColors[0], gammaCorrectedScaledColors[1], gammaCorrectedScaledColors[2]);
}

function sumRGBContributions(gammaCorrectedRed: number, gammaCorrectedGreen: number, gammaCorrectedBlue: number) {
    const contributionsofRedChannel = COEFFICIENTS_OF_CIE_COLOR_SPACE.RED * gammaCorrectedRed;
    const contributionsofGreenChannel = COEFFICIENTS_OF_CIE_COLOR_SPACE.GREEN * gammaCorrectedGreen;
    const contributionsofBlueChannel = COEFFICIENTS_OF_CIE_COLOR_SPACE.BLUE * gammaCorrectedBlue;

    return contributionsofRedChannel + contributionsofGreenChannel + contributionsofBlueChannel;
}

// See https://www.w3.org/TR/2008/REC-WCAG20-20081211/#contrast-ratiodef
export function calculateContrastRatio(backgroundRGB: number[], foregroundRGB: number[]) {
    const relativeLuminanceOfBackground = calculateRelativeLuminance(backgroundRGB[0], backgroundRGB[1], backgroundRGB[2]);
    const relativeLuminanceOfForeground = calculateRelativeLuminance(foregroundRGB[0], foregroundRGB[1], foregroundRGB[2]);

    // Adds small adjustment to avoid divide by zero errors.
    const greatestOfTwoRelativeLuminance = Math.max(relativeLuminanceOfBackground, relativeLuminanceOfForeground) + 0.05;
    const smalledOfTwoRelativeLuminance = Math.min(relativeLuminanceOfBackground, relativeLuminanceOfForeground) + 0.05;

    return (greatestOfTwoRelativeLuminance) / (smalledOfTwoRelativeLuminance);
}
