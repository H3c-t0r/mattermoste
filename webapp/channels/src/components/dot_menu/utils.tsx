// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {trackEvent} from 'actions/telemetry_actions';

import {EventTypes, TELEMETRY_CATEGORIES} from 'utils/constants';

import type React from 'react';

export type ChangeEvent = React.KeyboardEvent | React.MouseEvent;

export const trackDotMenuEvent = (e: ChangeEvent, suffix: string): void => {
    if (e.type === EventTypes.CLICK) {
        trackEvent(TELEMETRY_CATEGORIES.POST_INFO_MORE, EventTypes.CLICK + '_' + suffix);
    } else {
        trackEvent(TELEMETRY_CATEGORIES.POST_INFO_MORE, EventTypes.SHORTCUT + '_ ' + suffix);
    }
};
