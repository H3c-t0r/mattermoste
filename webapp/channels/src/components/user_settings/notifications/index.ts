// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {connect} from 'react-redux';
import {ActionCreatorsMapObject, bindActionCreators, Dispatch} from 'redux';

import {updateMe} from 'mattermost-redux/actions/users';
import {getConfig} from 'mattermost-redux/selectors/entities/general';
import {isCollapsedThreadsEnabled} from 'mattermost-redux/selectors/entities/preferences';
import {ActionFunc} from 'mattermost-redux/types/actions';

import {GlobalState} from 'types/store';

import UserSettingsNotifications, {Props} from './user_settings_notifications';
import {suitePluginIds} from 'utils/constants';
import semver from 'semver';

function mapStateToProps(state: GlobalState) {
    const config = getConfig(state);

    const sendPushNotifications = config.SendPushNotifications === 'true';
    const enableAutoResponder = config.ExperimentalEnableAutomaticReplies === 'true';
    const isCallsEnabled = semver.gte(state.plugins.plugins[suitePluginIds.calls].version, '0.17.0');

    return {
        sendPushNotifications,
        enableAutoResponder,
        isCollapsedThreadsEnabled: isCollapsedThreadsEnabled(state),
        isCallsEnabled,
    };
}

function mapDispatchToProps(dispatch: Dispatch) {
    return {
        actions: bindActionCreators<ActionCreatorsMapObject<ActionFunc>, Props['actions']>({
            updateMe,
        }, dispatch),
    };
}

export default connect(mapStateToProps, mapDispatchToProps)(UserSettingsNotifications);
