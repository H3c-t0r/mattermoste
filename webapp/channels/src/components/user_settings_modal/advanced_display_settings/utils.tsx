// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import React from 'react';
import {defineMessage, FormattedMessage} from 'react-intl';

import type {FieldsetCheckbox} from 'components/widgets/modals/components/checkbox_setting_item';
import type {FieldsetRadio} from 'components/widgets/modals/components/radio_setting_item';

import {isMac} from 'utils/user_agent';

export enum AdvanceSettings {
    CTRL_SEND='ctrlSend',
    SEND_ON_CTRL_ENTER= 'send_on_ctrl_enter',
    CODE_BLOCK_CTRL_ENTER= 'code_block_ctrl_enter',
    FORMATTING='formatting',
    JOIN_LEAVE='join_leave',
    NAME_DISABLE_CLIENT_PLUGINS= 'disable_client_plugins',
    NAME_DISABLE_TELEMETRY= 'disable_telemetry',
    NAME_DISABLE_TYPING_MESSAGES= 'disable_typing_messages',
    UNREAD_SCROLL_POSITION= 'unread_scroll_position',
    SYNCED_DRAFTS_ARE_ALLOWED = 'synced_drafts_are_allowed'
}

export const enum OnAndOff {
    ON= 'on',
    OFF= 'off',
}

export const enum UnreadScrollPosition {
    'LEFT'= 'start_from_left_off',
    'NEWEST'= 'start_from_newest',
}

export const getCtrlSelectedOption = (sendOnCtrlEnter: string, codeBlockOnCtrlEnter: string) => {
    if (sendOnCtrlEnter === 'true' && codeBlockOnCtrlEnter === 'false') {
        return AdvanceSettings.SEND_ON_CTRL_ENTER;
    } else if (sendOnCtrlEnter === 'false' && codeBlockOnCtrlEnter === 'true') {
        return AdvanceSettings.CODE_BLOCK_CTRL_ENTER;
    }
    return OnAndOff.OFF;
};

export const getCtrlSendPreferenceValue = (value: string): {[AdvanceSettings.SEND_ON_CTRL_ENTER]: string;[AdvanceSettings.CODE_BLOCK_CTRL_ENTER]: string} => {
    if (value === AdvanceSettings.SEND_ON_CTRL_ENTER) {
        return {[AdvanceSettings.SEND_ON_CTRL_ENTER]: 'true', [AdvanceSettings.CODE_BLOCK_CTRL_ENTER]: 'false'};
    } else if (value === AdvanceSettings.CODE_BLOCK_CTRL_ENTER) {
        return {[AdvanceSettings.SEND_ON_CTRL_ENTER]: 'false', [AdvanceSettings.CODE_BLOCK_CTRL_ENTER]: 'true'};
    }
    return {[AdvanceSettings.SEND_ON_CTRL_ENTER]: 'false', [AdvanceSettings.CODE_BLOCK_CTRL_ENTER]: 'false'};
};

export const ctrlSendInputFieldData: FieldsetRadio = {
    options: [
        {
            dataTestId: `${AdvanceSettings.CTRL_SEND}-${AdvanceSettings.SEND_ON_CTRL_ENTER}`,
            title: (
                <FormattedMessage
                    id='user.settings.advance.onForAllMessages'
                    defaultMessage='On for all messages'
                />
            ),
            name: `${AdvanceSettings.CTRL_SEND}`,
            key: `${AdvanceSettings.CTRL_SEND}-${AdvanceSettings.SEND_ON_CTRL_ENTER}`,
            value: AdvanceSettings.SEND_ON_CTRL_ENTER,
        },
        {
            dataTestId: `${AdvanceSettings.CTRL_SEND}-${AdvanceSettings.CODE_BLOCK_CTRL_ENTER}`,
            title: (
                <FormattedMessage
                    id='user.settings.advance.onForCode'
                    defaultMessage='On only for code blocks starting with ```'
                />
            ),
            name: `${AdvanceSettings.CTRL_SEND}`,
            key: `${AdvanceSettings.CTRL_SEND}-${AdvanceSettings.CODE_BLOCK_CTRL_ENTER}`,
            value: AdvanceSettings.CODE_BLOCK_CTRL_ENTER,
        },
        {
            dataTestId: `${AdvanceSettings.CTRL_SEND}-${OnAndOff.OFF}`,
            title: (
                <FormattedMessage
                    id='user.settings.advance.off'
                    defaultMessage='Off'
                />
            ),
            name: `${AdvanceSettings.CTRL_SEND}`,
            key: `${AdvanceSettings.CTRL_SEND}-${OnAndOff.OFF}`,
            value: OnAndOff.OFF,
        },
    ],
};

export const FormattingInputFieldData: FieldsetRadio = {
    options: [
        {
            dataTestId: `${AdvanceSettings.FORMATTING}-${OnAndOff.ON}`,
            title: (
                <FormattedMessage
                    id='ser.settings.advance.on'
                    defaultMessage='On'
                />
            ),
            name: `${AdvanceSettings.FORMATTING}`,
            key: `${AdvanceSettings.FORMATTING}-${OnAndOff.ON}`,
            value: OnAndOff.ON,
        },
        {
            dataTestId: `${AdvanceSettings.FORMATTING}-${OnAndOff.OFF}`,
            title: (
                <FormattedMessage
                    id='user.settings.advance.off'
                    defaultMessage='Off'
                />
            ),
            name: `${AdvanceSettings.FORMATTING}`,
            key: `${AdvanceSettings.FORMATTING}-${OnAndOff.OFF}`,
            value: OnAndOff.OFF,
        },
    ],
};

export const JoinLeaveInputFieldData: FieldsetRadio = {
    options: [
        {
            dataTestId: `${AdvanceSettings.JOIN_LEAVE}-${OnAndOff.ON}`,
            title: (
                <FormattedMessage
                    id='user.settings.advance.on'
                    defaultMessage='On'
                />
            ),
            name: `${AdvanceSettings.JOIN_LEAVE}`,
            key: `${AdvanceSettings.JOIN_LEAVE}-${OnAndOff.ON}`,
            value: OnAndOff.ON,
        },
        {
            dataTestId: `${AdvanceSettings.JOIN_LEAVE}-${OnAndOff.OFF}`,
            title: (
                <FormattedMessage
                    id='user.settings.advance.off'
                    defaultMessage='Off'
                />
            ),
            name: `${AdvanceSettings.JOIN_LEAVE}`,
            key: `${AdvanceSettings.JOIN_LEAVE}-${OnAndOff.OFF}`,
            value: OnAndOff.OFF,
        },
    ],
};

export const SyncedDraftInputFieldData: FieldsetRadio = {
    options: [
        {
            dataTestId: `${AdvanceSettings.SYNCED_DRAFTS_ARE_ALLOWED}-${OnAndOff.ON}`,
            title: (
                <FormattedMessage
                    id='user.settings.advance.on'
                    defaultMessage='On'
                />
            ),
            name: `${AdvanceSettings.SYNCED_DRAFTS_ARE_ALLOWED}`,
            key: `${AdvanceSettings.SYNCED_DRAFTS_ARE_ALLOWED}-${OnAndOff.ON}`,
            value: OnAndOff.ON,
        },
        {
            dataTestId: `${AdvanceSettings.SYNCED_DRAFTS_ARE_ALLOWED}-${OnAndOff.OFF}`,
            title: (
                <FormattedMessage
                    id='user.settings.advance.off'
                    defaultMessage='Off'
                />
            ),
            name: `${AdvanceSettings.SYNCED_DRAFTS_ARE_ALLOWED}`,
            key: `${AdvanceSettings.SYNCED_DRAFTS_ARE_ALLOWED}-${OnAndOff.OFF}`,
            value: OnAndOff.OFF,
        },
    ],
};

export const UnreadScrollPositionInputFieldData: FieldsetRadio = {
    options: [
        {
            dataTestId: `${AdvanceSettings.UNREAD_SCROLL_POSITION}-${OnAndOff.ON}`,
            title: (
                <FormattedMessage
                    id='user.settings.advance.startFromLeftOff'
                    defaultMessage='Start me where I left off'
                />
            ),
            name: `${AdvanceSettings.UNREAD_SCROLL_POSITION}`,
            key: `${AdvanceSettings.UNREAD_SCROLL_POSITION}-${OnAndOff.ON}`,
            value: UnreadScrollPosition.LEFT,
        },
        {
            dataTestId: `${AdvanceSettings.UNREAD_SCROLL_POSITION}-${OnAndOff.OFF}`,
            title: (
                <FormattedMessage
                    id='user.settings.advance.startFromNewest'
                    defaultMessage='Start me at the newest message'
                />
            ),
            name: `${AdvanceSettings.UNREAD_SCROLL_POSITION}`,
            key: `${AdvanceSettings.UNREAD_SCROLL_POSITION}-${OnAndOff.OFF}`,
            value: UnreadScrollPosition.NEWEST,
        },
    ],
};

export const getCtrlSendText = () => {
    const description = {
        default: defineMessage({
            id: 'user.settings.advance.sendDesc',
            defaultMessage: 'When enabled, CTRL + ENTER will send the message and ENTER inserts a new line.',
        }),
        mac: defineMessage({
            id: 'user.settings.advance.sendDesc.mac',
            defaultMessage: 'When enabled, ⌘ + ENTER will send the message and ENTER inserts a new line.',
        }),
    };
    const title = {
        default: defineMessage({
            id: 'user.settings.advance.sendTitle',
            defaultMessage: 'Send Messages on CTRL+ENTER',
        }),
        mac: ({
            id: 'user.settings.advance.sendTitle.mac',
            defaultMessage: 'Send Messages on ⌘+ENTER',
        }),
    };
    if (isMac()) {
        return {
            ctrlSendTitle: title.mac,
            ctrlSendDesc: description.mac,
        };
    }
    return {
        ctrlSendTitle: title.default,
        ctrlSendDesc: description.default,
    };
};

export const FormattingSectionTitle = defineMessage({
    id: 'user.settings.advance.formattingTitle',
    defaultMessage: 'Enable Post Formatting',
});

export const FormattingSectionDesc = defineMessage({
    id: 'user.settings.advance.formattingDesc',
    defaultMessage: 'If enabled, posts will be formatted to create links, show emoji, style the text, and add line breaks. By default, this setting is enabled.',
});

export const JoinLeaveSectionTitle = defineMessage({
    id: 'user.settings.advance.joinLeaveTitle',
    defaultMessage: 'Join/leave messages',
});

export const JoinLeaveSectionDesc = defineMessage({
    id: 'user.settings.advance.joinLeaveDesc',
    defaultMessage: 'When enabled, system messages display when users join or leave Mattermost channels.',
});

export const UnreadScrollPositionSectionTitle = defineMessage({
    id: 'user.settings.advance.unreadScrollPositionTitle',
    defaultMessage: 'Scroll position when viewing an unread channel',
});

export const UnreadScrollPositionSectionDesc = defineMessage({
    id: 'user.settings.advance.unreadScrollPositionDesc',
    defaultMessage: 'Choose your scroll position when you view an unread channel. Channels will always be marked as read when viewed.',
});

export const SyncedDraftSectionTitle = defineMessage({
    id: 'user.settings.advance.syncDrafts.Title',
    defaultMessage: 'Allow message drafts to sync with the server',
});

export const SyncedDraftSectionDesc = defineMessage({
    id: 'user.settings.advance.syncDrafts.Desc',
    defaultMessage: 'When enabled, message drafts are synced with the server so they can be accessed from any device. When disabled, message drafts are only saved locally on the device where they are composed.',
});

export const PerformanceDebuggingSectionTitle = defineMessage({
    id: 'user.settings.advance.performance.title',
    defaultMessage: 'Performance Debugging',
});

export const PerformanceDebuggingSectionDesc = defineMessage({
    id: 'user.settings.advance.performance.info1',
    defaultMessage: 'You may enable these settings temporarily to help isolate performance issues while debugging. We don\'t recommend leaving these settings enabled for an extended period of time as they can negatively impact your user experience. You may need to refresh the page before these settings take effect.',
});

export const DebuggingPluginInputFieldData: FieldsetCheckbox = {
    name: 'plugin',
    dataTestId: 'plugin',
};

export const DebuggingTelemetryInputFieldData: FieldsetCheckbox = {

    name: 'telemetry',
    dataTestId: 'telemetry',
};

export const DebuggingTypingInputFieldData: FieldsetCheckbox = {
    name: 'typing',
    dataTestId: 'typing',
};
