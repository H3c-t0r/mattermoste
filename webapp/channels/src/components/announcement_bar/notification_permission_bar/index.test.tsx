// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
// See LICENSE.txt for license information.

import {screen, fireEvent, act} from '@testing-library/react';
import React from 'react';

import {renderWithContext} from 'tests/react_testing_utils';
import {requestNotificationPermission, isNotificationAPISupported} from 'utils/notifications';

import NotificationPermissionBar from './index';

jest.mock('utils/notifications', () => ({
    requestNotificationPermission: jest.fn(),
    isNotificationAPISupported: jest.fn(),
}));

describe('NotificationPermissionBar', () => {
    const initialState = {
        entities: {
            users: {
                currentUserId: 'user-id',
            },
            general: {
                config: {},
                license: {},
            },
        },
    };

    beforeEach(() => {
        (isNotificationAPISupported as jest.Mock).mockReturnValue(true);
        (window as any).Notification = {permission: 'default'};
    });

    afterEach(() => {
        jest.clearAllMocks();
        delete (window as any).Notification;
    });

    test('should render the notification bar when conditions are met', () => {
        const {container} = renderWithContext(<NotificationPermissionBar/>, initialState);

        expect(container).toMatchSnapshot();
        expect(screen.getByText('We need your permission to show desktop notifications.')).toBeInTheDocument();
        expect(screen.getByText('Enable notifications')).toBeInTheDocument();
    });

    test('should not render the notification bar if user is not logged in', () => {
        const {container} = renderWithContext(<NotificationPermissionBar/>);

        expect(container).toMatchSnapshot();
        expect(screen.queryByText('We need your permission to show desktop notifications.')).not.toBeInTheDocument();
    });

    test('should not render the notification bar if Notifications are not supported', () => {
        delete (window as any).Notification;
        (isNotificationAPISupported as jest.Mock).mockReturnValue(false);

        const {container} = renderWithContext(<NotificationPermissionBar/>, initialState);

        expect(container).toMatchSnapshot();
        expect(screen.queryByText('We need your permission to show desktop notifications.')).not.toBeInTheDocument();
    });

    test('should call requestNotificationPermission and hide the bar when the button is clicked', async () => {
        (requestNotificationPermission as jest.Mock).mockResolvedValue('granted');

        renderWithContext(<NotificationPermissionBar/>, initialState);

        expect(screen.getByText('We need your permission to show desktop notifications.')).toBeInTheDocument();

        await act(async () => {
            fireEvent.click(screen.getByText('Enable notifications'));
        });

        expect(requestNotificationPermission).toHaveBeenCalled();
        expect(screen.queryByText('We need your permission to show desktop notifications.')).not.toBeInTheDocument();
    });
});
