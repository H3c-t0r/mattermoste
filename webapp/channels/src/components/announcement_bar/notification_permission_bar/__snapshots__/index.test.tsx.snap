// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NotificationPermissionBar should not render the notification bar if Notifications are not supported 1`] = `<div />`;

exports[`NotificationPermissionBar should not render the notification bar if user is not logged in 1`] = `<div />`;

exports[`NotificationPermissionBar should render the notification bar when conditions are met 1`] = `
<div>
  <div
    class="_StyledDiv-BRlth ksRYxX announcement-bar"
  >
    <div
      class="announcement-bar__text"
    >
      <i
        class="icon icon-alert-circle-outline"
      />
      <span>
        We need your permission to show desktop notifications.
      </span>
      <button>
        Enable notifications
      </button>
    </div>
    <a
      class="announcement-bar__close"
      href="#"
    >
      Ã—
    </a>
  </div>
</div>
`;
